core/src/main/scala/kafka/server/BrokerLifecycleManager.scala

```diff
@@ -0,0 +1,370 @@
+      Some(response.authenticationException())
+    } else if (response.versionMismatch() != null) {
+      Some(response.versionMismatch())
+    } else if (response.wasDisconnected()) {
+      Some(new IOException("Client was disconnected"))
+    } else if (!response.hasResponse) {
+      Some(new IOException("No response found"))
+    } else {
+      None
+    }
+  }
+
+  /**
+   * Task loop to be scheduled
+   *   - Schedule another heartbeat if no heartbeat is pending/in-flight AND time since last heartbeat >=
+   *     registration.heartbeat.interval.ms
+   *   - If time since last heartbeat > registration.lease.timeout.ms, FENCE the broker
+   *
+   */
+  private def processHeartbeatRequests(): Unit = {
+    // TODO: Do we want to allow some sort preemption to prioritize critical state changes?
+    // TODO: Ensure RPC timeout < heartbeat interval timeout (or at the very least < registration lease timeout)
+
+    // Ensure that there are no outstanding state change requests
+    if (pendingHeartbeat.compareAndSet(false, true)) {
+      // No pending heartbeat in-flight. We have a few things to check during this iteration
+      // Check when the last heartbeat was successful
+      // - If < registration.heartbeat.interval.ms, no-op
+      // - Else, check for any pending state changes that have been queued
+      //   - Attempt a heartbeat w/ the requested state change
+      // - No state change requests queued
+      //   - If > registration.lease.timeout.ms (We have fallen way behind and it's best to fence ourselves here)
+      //     - Fence ourselves and attempt a state change from FENCED -> ACTIVE in the next iteration
+      //   - If > registration.heartbeat.interval.ms
+      //     - Attempt another heartbeat w/ targetState = currentState
+      //
+      val timeSinceLastHeartbeat = TimeUnit.NANOSECONDS.toMillis(time.nanoseconds - lastSuccessfulHeartbeatTime)
+      // NOTE: We still have to ensure the last heartbeat was sent at least w/ a gap of registration.heartbeat.interval.ms
+      //       even though the task is scheduled at intervals registration.heartbeat.interval.ms because of
+      //       scheduler ticks being batched in some cases where another task hogs the scheduler's runtime.
+      //       We're not real-time here and so this accounts for two task runs occurring almost immediately one after
+      //       the other
+      if (timeSinceLastHeartbeat < config.registrationHeartbeatIntervalMs) {
+        // No-op
+        pendingHeartbeat.compareAndSet(true, false)
+        return
+      }
+
+      // Check for any pending state changes that have been queued
+      var state = requestQueue.poll
+      if (state == null) {
+        // No state change requests queued
+        if (timeSinceLastHeartbeat > config.registrationLeaseTimeoutMs) {
+          error(s"Last successful heartbeat was $timeSinceLastHeartbeat ms ago")
+          // Fence ourselves
+          currentState = metadata.BrokerState.FENCED
```

Comment:
The heartbeat needs to send a fence-local-broker event to the broker's metadata listener via something like this:

```
kafkaBroker.brokerMetadataListener.put(OutOfBandFenceLocalBrokerEvent(brokerEpoch))
```